### PR TITLE
feat: extend twin

### PR DESCRIPTION
**Pull request by DylanVerstraete**

### CODE REVIEW

**Comment by DylanVerstraete** [comment]
Need to update image image in order to let CI pass

File: twin.go [hunk 1]
```diff
@@ -96,13 +137,22 @@ func (s *Substrate) GetTwin(id uint32) (*Twin, error) {
 }
 
 // CreateTwin creates a twin
-func (s *Substrate) CreateTwin(identity Identity, ip net.IP) (uint32, error) {
+func (s *Substrate) CreateTwin(identity Identity, relay *string, pk []byte) (uint32, error) {
```
Comment (xmonader) [inline twin.go:140]: please make it `relay string` and use pointer semantics later in the code.

File: twin.go [hunk 2]
```diff
@@ -96,13 +137,22 @@ func (s *Substrate) GetTwin(id uint32) (*Twin, error) {
...
 	cl, meta, err := s.getClient()
 	if err != nil {
 		return 0, err
 	}
 
-	c, err := types.NewCall(meta, "TfgridModule.create_twin", ip.String())
+	relayOption := OptionRelay{}
+	pkOption := types.NewOptionBytesEmpty()
+	if relay != nil {
+		relayOption = OptionRelay{HasValue: true, AsValue: *relay}
```
Comment (xmonader) [inline twin.go:149]: after conversion to regular value, use the pointer semantics `AsValue: &relay`

File: twin.go [hunk 3]
```diff
@@ -96,13 +137,22 @@ func (s *Substrate) GetTwin(id uint32) (*Twin, error) {
...
+	}
+	if pk != nil {
```
Comment (xmonader) [inline twin.go:151]: the code should return an error if pk == nil in the very beginning of the function

File: twin.go [hunk 4]
```diff
@@ -115,13 +165,23 @@ func (s *Substrate) CreateTwin(identity Identity, ip net.IP) (uint32, error) {
 }
 
 // UpdateTwin updates a twin
-func (s *Substrate) UpdateTwin(identity Identity, ip net.IP) (uint32, error) {
+func (s *Substrate) UpdateTwin(identity Identity, relay *string, pk *string) (uint32, error) {
```
Comment (xmonader) [inline twin.go:168]: please change all to string and use pointer semantics later in the function body

File: twin.go [hunk 5]
```diff
@@ -115,13 +165,23 @@ func (s *Substrate) CreateTwin(identity Identity, ip net.IP) (uint32, error) {
...
 	cl, meta, err := s.getClient()
 	if err != nil {
 		return 0, err
 	}
 
-	c, err := types.NewCall(meta, "TfgridModule.update_twin", ip.String())
+	relay_bytes := types.OptionBytes{}
+	if relay != nil {
```
Comment (xmonader) [inline twin.go:175]: improve the validation on the relay and pk in the very top of the function please and terminate early with proper errors.

File: twin.go (hunk 3 above)
Comment (DylanVerstraete) [reply twin.go:151]: Both PK and relay are optional values, no need to return an error if it's not passed

**Comment by DylanVerstraete** [comment]
@muhamadazmy can merge?

**Review by muhamadazmy — APPROVED** [review]

File: twin.go [hunk 6]
```diff
@@ -115,13 +166,23 @@ func (s *Substrate) CreateTwin(identity Identity, ip net.IP) (uint32, error) {
 }
 
 // UpdateTwin updates a twin
-func (s *Substrate) UpdateTwin(identity Identity, ip net.IP) (uint32, error) {
+func (s *Substrate) UpdateTwin(identity Identity, relay string, pk string) (uint32, error) {
```
Comment (muhamadazmy) [inline twin.go:169]: Pk should not be a string instead a `[]byte`

**Review by muhamadazmy — CHANGES_REQUESTED** [review]

**Review by muhamadazmy — APPROVED** [review]

**Pull request merged**